File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/model/AemImport.java
package org.cru.godtools.articles.aem.model;

import android.arch.persistence.room.Entity;
import android.arch.persistence.room.PrimaryKey;
import android.net.Uri;
import android.support.annotation.NonNull;

import java.util.Date;

import static org.ccci.gto.android.common.base.TimeConstants.DAY_IN_MS;

@Entity(tableName = "aemImports")
public class AemImport {
    private static final long STALE_AGE = DAY_IN_MS;

    @NonNull
    @PrimaryKey
    public final Uri uri;

    @NonNull
    public Date lastProcessed = new Date(0);

    public AemImport(@NonNull final Uri uri) {
        this.uri = uri;
    }

    public boolean isStale() {
        return lastProcessed.before(new Date(System.currentTimeMillis() - STALE_AGE));
    }
}

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/model/AemImport.kt
package org.cru.godtools.articles.aem.model

import android.net.Uri
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import org.ccci.gto.android.common.base.TimeConstants.DAY_IN_MS
import java.util.Date

private const val STALE_AGE = DAY_IN_MS

@Entity(tableName = "aemImports")
class AemImport(@field:PrimaryKey val uri: Uri) {
    var lastProcessed = Date(0)

    fun isStale(): Boolean {
        return lastProcessed.before(Date(System.currentTimeMillis() - STALE_AGE))
    }

    @Entity(tableName = "aemImportArticles",
            primaryKeys = ["aemImportUri", "articleUri"],
            indices = [Index("articleUri")],
            foreignKeys = [
                ForeignKey(entity = AemImport::class,
                        onUpdate = ForeignKey.RESTRICT, onDelete = ForeignKey.CASCADE,
                        parentColumns = ["uri"], childColumns = ["aemImportUri"]),
                ForeignKey(entity = Article::class,
                        onUpdate = ForeignKey.RESTRICT, onDelete = ForeignKey.CASCADE,
                        parentColumns = ["uri"], childColumns = ["articleUri"])])
    class AemImportArticle(val aemImportUri: Uri, val articleUri: Uri) {
        constructor(aemImport: AemImport, article: Article) : this(aemImport.uri, article.uri)
    }
}
